tc/tc_util.c

```diff
@@ -192,6 +192,86 @@ static const struct rate_suffix {
 };
 
 
+int get_percent_rate(unsigned int *rate, const char *str, char *dev)
+{
+	long value;
+	char percent[strlen(dev) - 1], r_str[20];
+	const char *pos;
+	int ret;
+	double d, r;
+
+	if (read_prop(dev, "speed", &value))
```

Comment:
On my Linux system, reading the "speed" property doesn't work for all interfaces. For example, it doesn't work for my wireless interface or for an Ethernet bridge. Can you find a better way to get the device's speed limit?

Reply:
There's also no guarantee that the user gave an argument for `dev` when invoking `tc`, so you need to check to make sure it's not `NULL` before trying to use it.

Reply:
Hi Cody, I've tried searching for better ways but it seems the only other option is to make an `ioctl` call. Would that be ok?

Reply:
Hi Cody, for wireless interfaces, there is a different ioctl call but for all other wired interfaces - including bridges, which assuming is dependent on the speed of the interface that is connected to the bridge (and that has to be wired) - there is the `ethtool` ioctl call which can find the device speed (they have different ioctl calls because of the different device drivers being used for wired vs wireless connections). But given only the device name from the user we'd have to assume whether it is a wired or wireless connection. Given this, I was thinking the approach would be to search for the device name in the wired device list using `ethtool` ioctl call and if that call fails, search for the device in the wireless devices list, and if that fails, return `no such device found` .

Reply:
Because in this way we're not assuming whether the interface is wired or wireless but we'd still be searching for the right interface because every interface name should be unique.

Reply:
Let me know if this approach is ok. As of right now `tc` handles calls only pertaining to wired devices (and that's why I guess an error for wireless connections showed up), so the functionality for handling wireless devices needs its appropriate low level functionality to be added into `tc`, before the ioctl call can be made. So do let me know. Thanks.

Reply:
OK, thanks for researching that. Let's assume that the interface is wired, and keep the way that you're doing it for now. But if we're unable to read a valid speed, make sure an error is returned. I can't tell whether `read_prop()` will return an error if given the name of a wireless device, so include that in your testing to make sure.

Reply:
Ok, will do. At the moment,`read_prop()` does return an error but `get_precent_rate()` does not, will add that in. Thanks.